src/analyses/base.ml

```diff
@@ -1738,7 +1738,7 @@ struct
       (* if M.tracing then M.tracel "set" ~var:firstvar "new state1 %a" CPA.pretty nst; *)
       (* If the address was definite, then we just return it. If the address
        * was ambiguous, we have to join it with the initial state. *)
-      let nst = if AD.cardinal lval > 1 then { nst with cpa = CPA.join st.cpa nst.cpa } else nst in
+      let nst = if AD.cardinal lval > 1 then D.join st nst else nst in
```

Comment:
For ambiguous points-to sets, the join between old and new (with everything updated sequentially by the preceeding fold) is an optimization over executing each write separately and then joining all those new results (and not the old one).
But is that optimization even correct for (arbitrary) privatizations?

This certainly seems more sound than before, but I'm wondering if it's enough. Maybe we should just ditch the optimization and do the more naive thing that is obviously sound. All the potential writes are still done (either sequentially or in parallel), so this optimization only saves `AD.cardinal lval - 2` joins. Not sure if this is significant enough.

Reply:
I had similar doubts and so tried to replace the entire line here with:

~~~C
AD.fold (fun addr acc -> D.join (update_one addr st) acc) lval (D.bot ())
~~~

but that lead to some strange test failures where I really have no idea what's going on:

```
4 test(s) failed: ["04/35 trylock_rc", "04/36 trylock_nr", "41/05 more"]
```

If we want to clean this up, we have to investigate what happens with these.

I think for the privatizations it should not be a problem though, as all the may and must sets behave as one would expect.

Reply:
That's odd indeed. I'm fine with the current improvement then because it fixes the issue and doesn't break anything else.
The naive implementation failing some tests deserves an issue though. I had a quick tracing look and somehow there seemed to be `CPA` maps with many variables missing or `Uninitialized` (i.e. superbot) values, so something odd happens. Maybe when the address set is somehow empty, the new `D.bot ()` comes into play?

Reply:
Moved to #1465.